### ISSUE TITLE
Error message "Internal or recursive type is not allowed for public state variables" causes confusion

### ISSUE THREAD

**Issue by haoyang9804** [ethereum/solidity#15583]
## Description

The following code snippet causes an error with message `recursive type is not allowed for public state variables`. I cannot find references to "recursive type" in the doc https://docs.soliditylang.org/en/latest/types.html. Can we improve the doc or the error message

```solidity
contract contract0 {
  struct struct1 {
    mapping(int64 => int64)[10] array2;
  }
  mapping(int64 => struct1) public mapping11;
}
```
## Environment

- Compiler version: 0.8.29

**Comment by haoyang9804** [comment]
Another example:
```solidity
contract contract0 {
  struct struct1 {
    mapping(int256 => int256) mapping2;
    int256 var5;
  }

  struct struct2 {
    struct1 struct_instance3;
  }

  struct1 public struct_instance3;
  struct2 public struct_instance4;
}
```

The declaration of `struct_instance4` raises the same error.
@matheusaaguiar @ekpyron Do you think we should enrich the doc? I cannot find references in the doc and my colleague and I both think this error message is a little bit confusing.

**Comment by cameel** [comment]
Indeed, this message is not accurate. In this case the real cause are the nested mappings. What we really check for is that the type cannot be used in the external interface, but instead of having a generic message saying that, we try to be helpful and make assumptions about why that is. We don't verify those assumptions though.

Looks like we do distinguish these two cases when type checking errors and events: https://github.com/ethereum/solidity/blob/df91531c3aecf6477613874113a783bfb99ec6d0/libsolidity/analysis/TypeChecker.cpp#L4056-L4063

but not when type checking getters: https://github.com/ethereum/solidity/blob/df91531c3aecf6477613874113a783bfb99ec6d0/libsolidity/analysis/TypeChecker.cpp#L553-L556

Adding a similar check there would solve the problem. Wouldn't hurt to also add some asserts to ensure our assumptions are correct.

**Comment by floor-licker** [comment]
Currently working on this

**Comment by floor-licker** [comment]
@cameel I implemented the following changes to address the case of nested mappings but it doesn't seem to be detecting the nested mapping return type in the case haoyang9804 initially provides. I'm a bit stumped as to why that is.

The logic here is since we first assert that `getter.returnParameterTypes()` and `getter.parameterNames()`are not empty, we should then be able to iterate through those return types and check whether `paramType->containsNestedMapping()`. For some reason the condition `if (paramType->containsNestedMapping())` never catches haoyang9804's case.

```
if (!getter.interfaceFunctionType())
        {
            solAssert(getter.returnParameterNames().size() == getter.returnParameterTypes().size());
            solAssert(getter.parameterNames().size() == getter.parameterTypes().size());
            if (getter.returnParameterTypes().empty() && getter.parameterTypes().empty())
                m_errorReporter.typeError(5359_error, _variable.location(), "The struct has all its members omitted, therefore the getter cannot return any values.");
            
            for (auto const& paramType : getter.returnParameterTypes())
            {
                if (paramType->containsNestedMapping())
                {
                    m_errorReporter.typeError(
                        3448_error,
                        _variable.location(),
                        "Type containing a (nested) mapping is not allowed for public state variables."
                    );
                }
            }
            
            //Otherwise variable is Internal or recursive
            m_errorReporter.typeError(6744_error, _variable.location(), "Internal or recursive type is not allowed for public state variables.");
        }
```
